File: DSA/Linked_Lists/palindrome-linked-list.cpp
/*
https://leetcode.com/problems/palindrome-linked-list/
*/

#include <bits/stdc++.h> 
using namespace std; 

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
    
    ListNode *reverse_sll(ListNode *head) {
        
        ListNode *curr = head;
        ListNode *prev = NULL;
        ListNode *next;
        
        while (curr) {
            
            next = curr -> next;
            curr -> next = prev;
            prev = curr;
            curr = next;
        }
        
        return prev;
    }
    
    ListNode *get_palindrome_middle(ListNode *head) {
        
        ListNode *fast = head;
        ListNode *slow = head;       
        
        bool odd = true;
        
        while (fast && fast -> next) {
            
            fast = fast -> next -> next;
            slow = slow -> next;            
            
            if (fast == NULL)
                odd = false;
        }
        
        if (odd)
            slow = slow -> next;
        
        return slow;
        
    }
public:
    bool isPalindrome(ListNode* head) {
        
        if (head == NULL || head -> next == NULL)
            return true;
        
        ListNode *left = head;
        ListNode *right = get_palindrome_middle(head);
        
        right = reverse_sll(right);
        
        while (left && right) {
            
            if (left -> val != right -> val)
                return false;
            
            left = left -> next;
            right = right -> next;
        }
        
        return true;
        
    }

    ListNode* insert_list (int n) {

        if (n == 0)
            return NULL;
        
        int num;
        cin >> num;
        ListNode *root = new ListNode(num);
        ListNode *curr = root;
        ListNode *newnode;

        for (int i = 1; i < n; i++) {
            cin >> num;
            newnode = new ListNode(num);
            curr -> next = newnode;
            curr = curr -> next;
        }
        return root;
    }
};

int main() {

    /*
    5
    1 2 3 2 1
    */

    /*
    true
    */

    Solution *soln = new Solution;
    int n1;
    cin >> n1;

    ListNode *l1 = soln -> insert_list(n1);
    
    cout << soln -> isPalindrome(l1) << endl;

    return 0;

}    

File: Data-Structures-Algorithms/Linked-Lists/Singly-Linked-Lists/palindrome-linked-list.py
# https://leetcode.com/problems/palindrome-linked-list/submissions/1219383336/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        fast = head
        slow = head

        # find middle wit slow
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # rev 2nd half
        prev = None
        while slow:
            tmp = slow.next
            slow.next = prev
            prev = slow
            slow = tmp

        # check palindrome
        left = head
        right = prev

        while right:
            if left.val != right.val:
                return False
            left = left.next
            right = right.next
        
        return True

        


    def isPalindrome2(self, head: Optional[ListNode]) -> bool:
        stack = []

        curr = head
        while curr:
            stack.append(curr.val)
            curr = curr.next
        
        curr = head
        while (stack):
            rear = stack[-1]
            
            if (rear != curr.val):
                return False
            else:
                stack.pop()
                curr = curr.next
        
        return True


        